ast/compile.go

```diff
@@ -830,30 +856,64 @@ func astNodeToString(x interface{}) string {
 
 // checkRuleConflicts ensures that rules definitions are not in conflict.
 func (c *Compiler) checkRuleConflicts() {
+	rw := rewriteVarsInRef(c.RewrittenVars)
+
 	c.RuleTree.DepthFirst(func(node *TreeNode) bool {
 		if len(node.Values) == 0 {
-			return false
+			return false // go deeper
 		}
 
-		kinds := map[DocKind]struct{}{}
+		kinds := map[RuleKind]struct{}{}
 		defaultRules := 0
 		arities := map[int]struct{}{}
+		name := ""
+		var singleValueConflicts []Ref
 
 		for _, rule := range node.Values {
 			r := rule.(*Rule)
-			kinds[r.Head.DocKind()] = struct{}{}
+			ref := r.Ref()
+			name = rw(ref.Copy()).String() // varRewriter operates in-place
+			kinds[r.Head.RuleKind()] = struct{}{}
 			arities[len(r.Head.Args)] = struct{}{}
 			if r.Default {
 				defaultRules++
 			}
+
+			// Single-value rules may not have any other rules in their extent: these pairs are invalid:
```

Comment:
This is where the current limitation comes in: We're introducing string prefixes -- that limits the amount of potential overlaps between multiple rules.